proposed/non-primitive-constant-data.md

```diff
@@ -0,0 +1,471 @@
+# Non-primitive Constant Data
+
+## Introduction
+Current implementations of languages in .NET are limited to a small set of possible constant types. It would be useful for computational scenarios to provide a mechanism to specify efficiently accessible pre-computed data in more complex data structures. This proposal describes a mechanism for supporting user data structures of more arbitrary data structures, supporting both array production, and constant data access to a single constant as well as a ReadOnlySpan of data.
+
+## Current valid constant forms
+Currently accepted are 4 forms of constant
+- Integer/Float constants
+```csharp
+void Constants()
+{
+   int x = 3;
+   double y = 4.0;
+}
+```
+
+Which is represented in IL via the various ldc opcodes.
+```
+IL_0001: ldc.i4.s 3
+IL_0003: stloc.s 0
+IL_0005: ldc.r8 4.0
+IL_000E: stloc.s 1
+```
+
+- String constants
+```csharp
+void Constants()
+{
+    string z = "z";
+}
+```
+
+Which is represented in IL using the ldstr instruction and #US metadata table.
+```
+IL_0001: ldstr "z"
+```
+- Byte data constants
+```csharp
+void Constants()
+{
+    ReadOnlySpan<byte> byte_data = new ReadyOnlySpan<byte>(new byte[]{0,1,2,3});
+}
+```
+Which is represented in IL using an RVA static field on an anonymous type and the accessed via ldsflda to get its address.
+```
+IL_0001: ldloca.s 0
+IL_0003: ldsflda valuetype '<PrivateImplementationDetails>'/'__StaticArrayInitTypeSize=3' '<PrivateImplementationDetails>'::'0C7A623FD2BBC05B06423BE359E4021D36E721AD'
+IL_0008: ldc.i4.3
+IL_0009: call instance void valuetype [System.Memory]System.ReadOnlySpan`1<uint8>::.ctor(void*, int32)
+```
+
+
+- Primitive data for a dynamically constructed array
+```csharp
+void Constants()
+{
+    int[] data = new byte[]{0,1,2,4};
+}
+```
+
+Which is represented in IL using an RVA static field on an anonymous type which is used in conjunction with the System.Runtime.CompilerServices.RuntimeHelpers.InitializeArray function to dynamically create an array based on a set of data which was statically placed into the assembly.
+```
+IL_0001: ldc.i4.4
+IL_0002: newarr [mscorlib]System.Byte
+IL_0007: dup
+IL_0008: ldtoken field int32 '<PrivateImplementationDetails>'::'12DADA1FFF4D4787ADE3333147202C3B443E376F'
+IL_000d: call void [mscorlib]System.Runtime.CompilerServices.RuntimeHelpers::InitializeArray(class [mscorlib]System.Array, valuetype [mscorlib]System.RuntimeFieldHandle)
+IL_0012: stloc.0
+```
+## New valid constant forms
+
+The proposal is to provide a set of CompilerServices apis implemented as intrinsics (in most cases) which will allow the use of more complex constants. The intention is to permit more complex constant forms for ReadOnlySpan<T> access, constant access, and through IL compiler magic, add new forms of array initialization that are not currently supported through the InitializeArray pattern.
+
+### Definition of types which can be represented as constants
```

Comment:
The list of banned types should include mono's size-variadic types `nint` and `nfloat` as they don't fit on any of those categories.